lib/quicly.c

```diff
@@ -4375,15 +4681,184 @@ static int handle_ping_frame(quicly_conn_t *conn, struct st_quicly_handle_payloa
+         * sequence) */
+        if (schedule_retire_connection_id(conn, frame.sequence) != 0)
+            return QUICLY_TRANSPORT_ERROR_INTERNAL;
+        /* do not install this CID */
+        return 0;
+    }
+    /* First, handle retire_prior_to field.
+     * This order is important as it is possible to receive a NEW_CONNECTION_ID frame
+     * such that it retires active_connection_id_limit CIDs and then installs
+     * one new CID.
+     */
+    if (frame.retire_prior_to > conn->super.peer.max_retire_prior_to) {
+        if (frame.retire_prior_to > conn->super.peer.cid_sequence) {
+            if (schedule_retire_connection_id(conn, conn->super.peer.cid_sequence) != 0) {
+                /* TODO: if the scheduling fails (likely due to malloc failure),
+                 * it's possible that we could continue without disrupting the connection
+                 * -- just silently ignore retire_prior_to and the peer will eventually
+                 * retransmit the same information.
+                 * However given that malloc failure indicates catastrophic events are going
+                 * on in the system, probably it's reasonable to drop the connection by
+                 * returning the error. */
+                return QUICLY_TRANSPORT_ERROR_INTERNAL;
+            }
+            need_to_replace_cid = 1;
+        }
+
+        /* retire CIDs in the spare pool */
+        for (int i = 0; i < QUICLY_LOCAL_ACTIVE_CONNECTION_ID_LIMIT - 1; i++) {
+            struct st_quicly_spare_cid_t *spare_cid = &conn->super.peer.spare_cids[i];
+            if (!spare_cid->is_active || spare_cid->sequence >= frame.retire_prior_to)
+                continue;
+            if (schedule_retire_connection_id(conn, spare_cid->sequence) != 0)
+                return QUICLY_TRANSPORT_ERROR_INTERNAL;
+            spare_cid->is_active = 0;
+        }
+        conn->super.peer.max_retire_prior_to = frame.retire_prior_to;
+    }
+
+    if (need_to_replace_cid) {
+        struct st_quicly_spare_cid_t *spare_cid = pick_spare_cid(conn);
+        if (spare_cid == NULL) {
+            /* the peer has retired every CID and is installing a new one */
+            set_cid(&conn->super.peer.cid, frame.cid);
+            memcpy(conn->super.peer.stateless_reset._buf, frame.stateless_reset_token, QUICLY_STATELESS_RESET_TOKEN_LEN);
+            conn->super.peer.stateless_reset.token = conn->super.peer.stateless_reset._buf;
+            conn->super.peer.cid_sequence = frame.sequence;
+            return 0;
+        }
+        /* replace with the picked one */
+        spare_cid->is_active = 0;
+        conn->super.peer.cid = spare_cid->cid;
+        memcpy(conn->super.peer.stateless_reset._buf, spare_cid->stateless_reset_token, QUICLY_STATELESS_RESET_TOKEN_LEN);
+        conn->super.peer.stateless_reset.token = conn->super.peer.stateless_reset._buf;
+        conn->super.peer.cid_sequence = spare_cid->sequence;
+    }
```

Comment:
The logic here has two ways of picking a new CID. I wonder if we can merge the two into one, if we change the logic of this function to at first register the value carried by the NCID frame then pick one from the spare list.

Reply:
👍 Done.